### DISCUSSION TITLE
Unable to login using Twitter Provider

### DISCUSSION THREAD

**Discussion by hishamco** [OrchardCMS/OrchardCore#12110] [Q&A]
I did some refactoring on `OC.Twitter` module, when I enable the module and setup the twitter settings, then I go to Login I got the following error:

InvalidOperationException: An error has occurred while calling the Twitter API, error's returned:
Code: 415, Message: 'Callback URL not approved for this client application. Approved callback URLs can be adjusted in your application settings'

FYI I didn't set the callback in the settings

**Comment by ns8482e** [comment]
@hishamco Do as error suggests ;)

Add redirect url in Twitter app

**Reply by hishamco** [reply]
I already did, also one more thing I need to test this from localhost, the Twitter settings required a domain, I will try to handle this using hosts file

I just need to make sure that anyone tested before

**Comment by MichaelPetrinolis** [comment]
I did the twitter module and confirm that it was working back then ;)

**Reply by ns8482e** [reply]
It seems to be working - I'll validate

**Reply by hishamco** [reply]
@MichaelPetrinolis thanks, could you please let me know what kind of settings needed to make it runs on localhost?

**Reply by ns8482e** [reply]
I got this error 

An unhandled exception occurred while processing the request.
InvalidOperationException: An error has occurred while calling the Twitter API, error's returned:
Code: 453, Message: 'You currently have Essential access which includes access to Twitter API v2 endpoints only.

May be something changed in twitter api or i am might be missing some config/settings

**Reply by hishamco** [reply]
I forgot to point that there's a new AUTH v2 showed up in the settings, @ns8482e could you please show me what the settings that are using while there's no secrets here ;)

**Reply by MichaelPetrinolis** [reply]
> @MichaelPetrinolis thanks, could you please let me know what kind of settings needed to make it runs on localhost?

@hishamco I checked my twitter account and I had defined localhost as Callback URI/Redirect URL in the app settings. Maybe they disabled it, like FB did some time ago. Also only OAuth 1 was supported at the time of writing (which supports both Twitter APIs).

here is a screenshot of a test twitter app
![image](https://user-images.githubusercontent.com/9623255/182046497-d0a507a1-2d5d-48b3-a1fe-10d99c71dc45.png)

**Reply by ns8482e** [reply]
Seems V1 is not supported may be need to update to V2

**Reply by hishamco** [reply]
I will check both of them while I'm working on the Twitter module

**Reply by hishamco** [reply]
When I used V1 there are a strange loop happenning

singin-twitter with OAUTH token
ExternalLoginCallBack
Login
DefaultExternalLogin
authenticate with OAUTH token

When I used V2 it doesn't redirect, I presume we need to change the code to interact with the newer APIs

@MichaelPetrinolis could you please verify that's working with the latest bits?

**Reply by hishamco** [reply]
Side note, when you enable the twitter feature, this message showed app "Site does not allow user registration." Because there's a settings to be enables. I'm suggesting to mention this in the docs, to avoid the expectation

**Reply by hishamco** [reply]
I created another tenant, same loop happen. I can confirm something weird happening :(

**Reply by MichaelPetrinolis** [reply]
> I created another tenant, same loop happen. I can confirm something weird happening :(

@hishamco I will give it a try as soon as I return from vacations.

**Comment by MichaelPetrinolis** [comment]
@hishamco @ns8482e I tested it locally with the main branch and it works. I also created a tenant on https://try.orchardcore.net , and enabled twitter sign in. Configured twitter integration, and in my twitter app added the callback. 
It works as expected.

**Reply by hishamco** [reply]
Oh really!! are you using V1 or V2?

**Reply by MichaelPetrinolis** [reply]
V1. The Twitter Sign In module uses V1. Maybe we could add a feature to support V2 apps.
Do you want me to send you the link from try orchard core to test with your credentials ?

**Reply by hishamco** [reply]
It will be great, BTW which version that https://try.orchardcore.net/ is working on?

**Reply by MichaelPetrinolis** [reply]
I sent you the link. I don't know what version TryOrchardCore runs. @agriffard try orchard is on latest release ?
I also tried it locally on the latest bits and works.

**Reply by hishamco** [reply]
Where did you send it?

**Reply by MichaelPetrinolis** [reply]
To your email from Github profile

**Reply by hishamco** [reply]
Let me check my Yahoo mail coz I'm not check it from awhile ;)

Thanks

**Reply by hishamco** [reply]
@MichaelPetrinolis I didn't find it in hunderds of emails ;) could you please resend it to hishamco_2007@hotmail.com

**Reply by MichaelPetrinolis** [reply]
done

**Reply by hishamco** [reply]
Could you please remove your secrets, then I will try mine? Also you can create an admin user for me so I can enable to login

**Reply by MichaelPetrinolis** [reply]
sent you the admin credentials. I reset my twitter api secrets, so feel free to perform your tests!

**Reply by hishamco** [reply]
Thanks a lot

**Reply by hishamco** [reply]
I got "Invalid login attempt"

**Reply by MichaelPetrinolis** [reply]
Just tried it, it works. maybe you accidentally copied a space with the password? ( i mean the admin password)

**Reply by hishamco** [reply]
Ya extra space ;)

**Reply by hishamco** [reply]
When it redirects to https://try.orchardcore.net/noy0poe0/OrchardCore.Users/Account/ExternalLogin I'm getting 
**An error occurred while executing this request**

**Reply by MichaelPetrinolis** [reply]
Will retry to create a new twitter app and test.
could you regenerate your api keys to test?
this is my app config 
![image](https://user-images.githubusercontent.com/9623255/185642132-abd34965-04e2-44c8-81fa-ded49914b027.png)

**Reply by MichaelPetrinolis** [reply]
you must provide app privacy and terms of service URLs for the app also.
